### ISSUE TITLE
Unicode processing issue in terminal buffers

### ISSUE THREAD

**Issue by foonathan** [neovim/neovim#16245]
### Neovim version (nvim -v)

NVIM v0.5.1 (also present on HEAD)

### Vim (not Nvim) behaves the same?

n/a (vim doesn't have `termopen()`)

### Operating system/version

ArchLinux

### Terminal name/version

kitty 0.23.1-1 (also present with xterm and a TTY)

### $TERM environment variable

xterm-kitty

### Installation

pacman -S neovim

### How to reproduce the issue

Take the following C program that repeatedly prints a Unicode character with a pause in between:

```c
#include <stdio.h>

int main()
{
    // I haven't tested other Unicode characters.
    // The multilines and the initial whitespace seem to trigger it easier.
    const char str[] = "   │\n   │\n"; // not ASCII bar, but U+2502

    while (1)
    {
        // If I use fputs(), the error happens a lot less often.
        for (const char* ptr = str; *ptr; ++ptr)
            fputc(*ptr, stderr);

        // We need some sort of pause here to trigger it.
        int i;
        scanf("%d", &i);
    }
}
```

Compile it (e.g `gcc print.c`) and start nvim --clean. Enter `:call termopen('./a.out')` and go to insert/terminal mode. Repeatedly enter some number to execute ~10 loop iterations.

### Expected behavior

The terminal buffer contains output like this:

```
   │
   │
0
   │
   │
0
etc
```

### Actual behavior

The terminal buffer contains output like this:

```
   │
   │
0
   │
   │
0
   ��
�   │
0
   �
�   │
0
   │
   │
0
   │
   │
0
   │
   │
0
   │
   ��
�0
   │
   ��
�0
```

This output was created by copy-pasting the terminal buffer, so it is not a rendering issue - it actually contains invalid Unicode. Note that not every output is mangled, only some, and this can vary from execution to execution (sometimes it works completely fine, sometimes almost everything is mangled). It also persists when you use C++'s `u8` string literal, so it's not a Unicode encoding issue of the source file or something like that.

Given this behavior and the fact that using `fputs()` over `fputc()` fixes it, I suspect it is some sort of buffering issue where neovim receives partial UTF-8 input and concludes that it has an invalid code unit: when it then receives the rest of the code point, it thinks that this is another invalid code unit, instead of combining the two into a valid code point.

**Comment by zeertzjq** [comment]
I guess this is the problem?
<https://github.com/neovim/neovim/blob/1fdbd29dfa6366f8346693d0bf67f4f782ab0f32/src/nvim/channel.c#L612>

**Comment by erw7** [comment]
This is probably a duplicate of #12609. I've been trying to debug it for a while now, and finally found the cause. This is caused by a bug in libvterm. If the sequence is incomplete when returning from [`decode_utf8()`](https://github.com/neovim/libvterm/blob/65dbda3ed214f036ee799d18b2e693a833a0e591/src/encoding.c#L28), `data->bytes_remaining` will be assigned the number of bytes remaining. However, when processing the rest of the sequence, in the following part of [`on_text()`](https://github.com/neovim/libvterm/blob/65dbda3ed214f036ee799d18b2e693a833a0e591/src/state.c#L238), a different `VTermEncodingInstance` is assigned to encoding than the previous one. Therefore, it is not processed as a continuation sequence and becomes `UNICODE_INVALID`(�, U+FFFD).

```C
  VTermEncodingInstance *encoding =
    state->gsingle_set     ? &state->encoding[state->gsingle_set] :
    !(bytes[eaten] & 0x80) ? &state->encoding[state->gl_set] :
    state->vt->mode.utf8   ? &state->encoding_utf8 :
                             &state->encoding[state->gr_set];
```

I used `fprintf` to output the `on_text()` and `decode_utf8()` status and got the following log.

```
state->gsingle_set: 0
state->gl_set: 0
!(bytes[eaten] & 0x80): 1
state->vt->mode.utf8: 1
state->gr_set: 1
encoding: 0x14fc6e8
BEGIN UTF-8, data->bytes_remaining: 0
END UTF-8, data->bytes_remaining: 1
state->gsingle_set: 0
state->gl_set: 0
!(bytes[eaten] & 0x80): 0
state->vt->mode.utf8: 1
state->gr_set: 1
encoding: 0x14fc748
BEGIN UTF-8, data->bytes_remaining: 0
```

CC: @leonerd

**Comment by foonathan** [comment]
For reference, here is a minimal vterm example that triggers it.

```c
static int putglyph(VTermGlyphInfo *info, VTermPos pos, void * user)
{
    printf("%d: ", info->width);
    for (int i = 0; i < info->width; ++i)
        printf("%x", info->chars[i]);
    printf("\n");
    return 1;
}

static const VTermStateCallbacks callbacks  = {
    .putglyph = &putglyph
};

int main(int argc, char *argv[])
{
  VTerm *vt = vterm_new(25, 80);
  vterm_set_utf8(vt, 1);

  VTermState* state = vterm_obtain_state(vt);
  vterm_state_reset(state, 1); // this seems to be necessary to initialize the encodings
  vterm_state_set_callbacks(state, &callbacks, NULL);

  vterm_input_write(vt, "a\xC3", 2); // need one full code point before the partial one
  vterm_input_write(vt, "\xA4", 1);

  vterm_free(vt);

  return 0;
}
```

**Comment by foonathan** [comment]
The following patch fixes that issue:

```diff
diff --git a/src/state.c b/src/state.c
index 9108009..6467925 100644
--- a/src/state.c
+++ b/src/state.c
@@ -246,11 +246,16 @@ static int on_text(const char bytes[], size_t len, void *user)
   int npoints = 0;
   size_t eaten = 0;
 
-  VTermEncodingInstance *encoding =
-    state->gsingle_set     ? &state->encoding[state->gsingle_set] :
-    !(bytes[eaten] & 0x80) ? &state->encoding[state->gl_set] :
-    state->vt->mode.utf8   ? &state->encoding_utf8 :
-                             &state->encoding[state->gr_set];
+  VTermEncodingInstance *encoding;
+  if (state->vt->mode.utf8) {
+    /* If the mode has been set to UTF-8, we're ignoring all shift functions. */
+    encoding = &state->encoding_utf8;
+  } else {
+    /* Select the encoding according to the shift functions. */
+    encoding = state->gsingle_set     ? &state->encoding[state->gsingle_set] :
+               !(bytes[eaten] & 0x80) ? &state->encoding[state->gl_set]      :
+                                        &state->encoding[state->gr_set];
+  }
 
   (*encoding->enc->decode)(encoding->enc, encoding->data,
       codepoints, &npoints, state->gsingle_set ? 1 : len,
```

As @erw7 has figured out, the issue is due to the fact that a different `VTermEncodingInstance` is used for the continuation byte: as `!(bytes[eaten] & 0x80)` is `false`, it does not use `state->encoding[state->gl_set]` as for the initial bytes but `state->encoding_utf8`, which has no idea about the previous bytes.

It is my understanding that the `gl_set`/`gr_set`/`gsingle_set` are set by various control codes to change the encoding of the input. I can't imagine that anybody who enables UTF-8 for the terminal cares about that functionality, so my patch simply ignores them and uses the `encoding_utf8` member all the time.

I've built nvim with that patch applied to libvterm and my original issue is fixed as well.

**Comment by erw7** [comment]
Probably, as https://github.com/neovim/neovim/issues/16245#issuecomment-962591284 says, we can use mode.utf8 at the beginning of the condition without any problem in most cases. However, when I checked on other terminals (`xterm`, `
Alacritty`, `st`, `kitty`), `smacs` worked even in UTF-8 locales. Therefore, it may be better to fix it as in the following patch.


```diff
diff --git a/src/encoding.c b/src/encoding.c
index 434ac3f..44ea56e 100644
--- a/src/encoding.c
+++ b/src/encoding.c
@@ -30,6 +30,7 @@ static void decode_utf8(VTermEncoding *enc, void *data_,
                         const char bytes[], size_t *pos, size_t bytelen)
 {
   struct UTF8DecoderData *data = data_;
+  bool beginning_8bit = (bytes[*pos] & 0x80);
 
 #ifdef DEBUG_PRINT_UTF8
   printf("BEGIN UTF-8\n");
@@ -42,21 +43,37 @@ static void decode_utf8(VTermEncoding *enc, void *data_,
     printf(" pos=%zd c=%02x rem=%d\n", *pos, c, data->bytes_remaining);
 #endif
 
-    if(c < 0x20) // C0
+    if(c < 0x20) { // C0
+      if(data->bytes_remaining)
+        cp[(*cpi)++] = UNICODE_INVALID;
+
+      data->bytes_remaining = 0;
       return;
+    }
 
     else if(c >= 0x20 && c < 0x7f) {
       if(data->bytes_remaining)
         cp[(*cpi)++] = UNICODE_INVALID;
 
+      data->bytes_remaining = 0;
+      if (beginning_8bit)
+        return;
+
       cp[(*cpi)++] = c;
 #ifdef DEBUG_PRINT_UTF8
       printf(" UTF-8 char: U+%04x\n", c);
 #endif
+    }
+
+    else if(c == 0x7f) { // DEL
+      if(data->bytes_remaining)
+        cp[(*cpi)++] = UNICODE_INVALID;
+
       data->bytes_remaining = 0;
+      return;
     }
 
-    else if(c == 0x7f) // DEL
+    else if(!beginning_8bit)
       return;
 
     else if(c >= 0x80 && c < 0xc0) {
```

**Comment by foonathan** [comment]
Any update on this?

**Comment by clason** [comment]
Clearly not, or you would have been notified on your own issue.

**Comment by foonathan** [comment]
Let me rephrase my question: is there anything I can do to fix the issue?  @erw7 has proposed a fix that works and seems reasonable. What's blocking it from applying the fix? Can I help with that in any way?

**Comment by clason** [comment]
Well, a PR would help, including a test that fails without and passes with that patch. (It's one thing to make a change that fixes a single issue; it's another thing to make sure that it does not break any of a myriad other edge cases...)
